libparsec/crates/platform_mountpoint/src/unix/inode.rs

```diff
@@ -0,0 +1,218 @@
+        NonZeroU64::try_from(value as u64 + 1)
+            .map(Self)
+            .expect("Unreachable")
+    }
+}
+
+impl From<Inode> for usize {
+    /// Convert from inode to the stack's index
+    ///
+    /// -> `Inode` starts at `1` which is the root directory
+    fn from(value: Inode) -> Self {
+        u64::from(value.0) as usize - 1
+    }
+}
+
+impl TryFrom<u64> for Inode {
+    type Error = TryFromIntError;
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        NonZeroU64::try_from(value).map(Self)
+    }
+}
+
+impl From<Inode> for u64 {
+    fn from(value: Inode) -> Self {
+        u64::from(value.0)
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct InodeManager {
+    global_paths: RwLock<GlobalPaths>,
+    opened: Mutex<HashMap<PathBuf, (Counter, Inode)>>,
+}
+
+impl<T: MountpointInterface> FileSystemWrapper<T> {
+    pub(super) fn insert_path(&self, path: PathBuf) -> Inode {
+        let mut global_paths = self
+            .inode_manager
+            .global_paths
+            .write()
+            .expect("Mutex is poisoned");
+        let mut opened = self.inode_manager.opened.lock().expect("Mutex is poisoned");
+
+        if let Some((counter, ino)) = opened.get_mut(&path) {
+            counter.increment();
+            return *ino;
+        }
+
+        let ino = if let Some(i) = global_paths.stack_unused_path.pop() {
+            global_paths.paths[i] = path.clone();
+            i
+        } else {
+            let i = global_paths.paths.len();
+            global_paths.paths.push(path.clone());
+            i
+        }
+        .into();
+
+        opened.insert(path, (Counter::default(), ino));
+        ino
+    }
+
+    /// # Safety:
+    /// It will panic if:
+    /// - `ino` does not exist
+    /// - `nlookup` is greater than the `counter` associated to the `ino`
+    pub(super) unsafe fn remove_path(&self, ino: Inode, nlookup: u64) {
+        let mut global_paths = self
+            .inode_manager
+            .global_paths
+            .write()
+            .expect("Mutex is poisoned");
+        let mut opened = self.inode_manager.opened.lock().expect("Mutex is poisoned");
+        let i: usize = ino.into();
+
+        let path = &global_paths.paths[i];
+
+        if let Some((counter, _)) = opened.get_mut(path) {
+            counter.decrement(nlookup);
+
+            if counter.is_zero() {
+                opened.remove(path);
+                global_paths.stack_unused_path.push(i);
+            }
+        }
+    }
+
+    /// Safety:
+    /// It will panic if:
+    /// - `ino` does not exist
+    pub(super) unsafe fn get_path(&self, ino: Inode) -> PathBuf {
+        let global_paths = self
+            .inode_manager
+            .global_paths
+            .read()
+            .expect("Mutex is poisoned");
+
+        global_paths.paths[usize::from(ino)].clone()
+    }
+
+    /// Safety:
+    /// It will panic if:
+    /// - `source` or `destination` are not `UTF-8` path
```

Comment:
we must handle non-utf8 path !

return an error in this case, typically file not found if source is invalid, and a more generic error if destination is invalid

Reply:
and should check this is also handled on winfsp